### PR TITLE
dev/financial#77 ++ Make contribution_id mandatory for PaymentProcessor.pay, pass invoiceID

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Replaces https://github.com/civicrm/civicrm-core/pull/15477 & also resolves https://lab.civicrm.org/dev/financial/issues/77

As discussed in Barcelona -  requiring contribution_id for PaymentProcessor.pay as a way to ensure that it is only called after the order is created
per our preferred flow (people could still get past this by passing a random number - but it feels like they would at least know they were hacking our process & take responsibility for any issues if it breaks or we suddenly start enforcing that
it is a valid contribution.

This also sets some of the recommended variables. 

Before
----------------------------------------
`contribution_id` not required by API PaymentProcessor.pay.

After
----------------------------------------
`contribution_id` required by API PaymentProcessor.pay.

Technical Details
----------------------------------------
See https://lab.civicrm.org/dev/financial/issues/53.

Technical Details
----------------------------------------
@mattwire it seems the test for https://lab.civicrm.org/dev/financial/issues/77 could easily also cover #15477 so I incorporated it.

Note that I had to use a few more lines to ensure we were always setting contactID, contributionRecurID to an integer. I do think this stricter approach is better but it wound up
more verbose

Comments
----------------------------------------

### CODE REVIEW

File: api/v3/PaymentProcessor.php [hunk 1]
```diff
@@ -124,11 +124,22 @@ function _civicrm_api3_payment_processor_getlist_defaults(&$request) {
...
  * @throws \API_Exception
+ * @throws \CiviCRM_API3_Exception
  */
 function civicrm_api3_payment_processor_pay($params) {
+  /* @var CRM_Core_Payment $processor */
   $processor = Civi\Payment\System::singleton()->getById($params['payment_processor_id']);
   $processor->setPaymentProcessor(civicrm_api3('PaymentProcessor', 'getsingle', ['id' => $params['payment_processor_id']]));
   try {
+    $processor->setContributionID($params['contribution_id']);
+    $processor->setInvoiceID($params['invoice_id'] ?? '');
```
Comment (mattwire) [inline api/v3/PaymentProcessor.php]: @eileenmcnaughton What did we agree in terms of invoice_id?  This means that we'll end up with an empty invoice ID if not passed in via params. Maybe that's ok!

Comment (eileenmcnaughton) [reply api/v3/PaymentProcessor.php]: @mattwire yep - it has to be passed in! We talked about Order.create adding one here https://lab.civicrm.org/dev/financial/issues/78 but haven't resolved that & although I marked it 'high' here https://lab.civicrm.org/dev/financial/issues/76 I'm re-thinking that & am probably not seeing it as something to worry about until some of the other things have been done

**Comment by mattwire** [comment]
@eileenmcnaughton Apart from the query on invoice_id this is good to merge

**Comment by eileenmcnaughton** [comment]
@mattwire FYI the invoice_id one shouldn't have a negative effect - it just means that it will be passed through if available

**Pull request merged**